File: web/src/utils/username.ts
/** Maximum number of ASCII characters in a writable username. */
export const MAX_USERNAME_LENGTH = 36;

const ASCII_ALPHANUMERIC = /^[A-Za-z0-9]$/;

/** Whether a character may occur anywhere in a writable username. */
export function isUsernameCharacter(char: string): boolean {
  return ASCII_ALPHANUMERIC.test(char) || char === "-";
}

/** Whether a value satisfies the writable username format. */
export function isValidUsername(username: string): boolean {
  if (username.length === 0 || username.length > MAX_USERNAME_LENGTH) return false;
  if (!ASCII_ALPHANUMERIC.test(username[0]) || !ASCII_ALPHANUMERIC.test(username.at(-1) ?? "")) return false;

  for (const char of username) {
    if (!isUsernameCharacter(char)) return false;
  }
  return true;
}

File: internal/base/username.go
package base

// MaxUsernameLength is the maximum number of ASCII characters in a writable username.
const MaxUsernameLength = 36

// IsValidUsername reports whether username satisfies the writable username format.
func IsValidUsername(username string) bool {
	if len(username) == 0 || len(username) > MaxUsernameLength || !isASCIIAlphanumeric(username[0]) || !isASCIIAlphanumeric(username[len(username)-1]) {
		return false
	}

	for i := 0; i < len(username); i++ {
		if !IsUsernameCharacter(username[i]) {
			return false
		}
	}
	return true
}

// IsUsernameCharacter reports whether an ASCII byte may occur in a writable username.
func IsUsernameCharacter(char byte) bool {
	return isASCIIAlphanumeric(char) || char == '-'
}

func isASCIIAlphanumeric(char byte) bool {
	return isASCIILetter(char) || isASCIIDigit(char)
}

func isASCIILetter(char byte) bool {
	return char >= 'A' && char <= 'Z' || char >= 'a' && char <= 'z'
}

func isASCIIDigit(char byte) bool {
	return char >= '0' && char <= '9'
}
